File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow,fast=head,head.next
        while fast and fast.next and slow!=fast:
            slow=slow.next
            fast=fast.next.next
        return True if slow==fast else False

File: Data Structures & Algorithms/non-cyclical-number/submission-2.java
class Solution {
    public boolean isHappy(int n) {
        int slow=n;
        int fast=happyCompute(happyCompute(n));
        while(slow!=fast){
            slow=happyCompute(slow);
            fast=happyCompute(happyCompute(fast));
        }
        return slow==1;
    
    }

    public int happyCompute(int n){
        int res=0;
        while (n>0){
            int remainder=n%10;
            res+=(remainder*remainder);
            n/=10;
        }
        return res;
    }
}
